core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1163,4 +1170,21 @@ public long getLastWriteNanos() {
      * This operation will close the engine if the recovery fails.
      */
     public abstract Engine recoverFromTranslog() throws IOException;
+
+    /**
+     * Called when a refresh includes the location.
+     */
+    public static interface RefreshListener {
+        /**
+         * The location to wait for. Must not return null and should be fast.
+         */
+        Translog.Location location();
```

Comment:
I had that and I didn't like that you'd end up building two classes - one for the runnable and one for the listener. I was actually thinking of making RefreshListener abstract and including the location but I didn't want to spend a ton more time on the PR until an initial review. I can certainly change it to the Location, Runnable version if you like it better. I'm not in love with any of it.